File: src/timesout.erl
-module(timesout).

-behaviour(gen_server).

-export([start_link/0]).
-export([yawn/1, catching_fun/1]).
-export([init/1, handle_call/3, handle_cast/2, terminate/2, code_change/3]).


start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, {}, []).

catching_fun(F) ->
    fun () ->
            try F() of
                Result ->
                    io:fwrite("Result: ~w~n", [Result])
            catch
                Err:Term ->
                    io:fwrite("Error occured ~w:~w~n", [Err, Term])
            end
    end.

yawn(Sleep) ->
    gen_server:call(?MODULE, {yawn, Sleep}, 100).

init(_) ->
    {ok, 0}.

handle_call({yawn, SleepTime}, _from, Count) ->
    timer:sleep(SleepTime),
    {reply, ok, Count + 1}.

handle_cast(_Msg, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

File: lib/timesout.ex
defmodule Timesout do
  @moduledoc """
  GenServer to help explore `GenServer.call/3` timeouts.

  To save hanging around, each call timesout after 100 millisconds
  """
  use GenServer



  @name __MODULE__
  @timeout 100

  def start_link() do
    GenServer.start_link(__MODULE__, {}, name: @name)
  end

  def init(_) do
    {:ok, 0}
  end

  @doc """
  Call the process and leeps for `sleep` milliseconds. If `sleep` > 100, or the
  GenServer is otherwise blocked, then the call should timeout. Returns the number of
  calls received by the GenServer, prior ot this one.
  """
  @spec yawn(integer) :: integer
  def yawn(sleep) do
    GenServer.call(@name, {:yawn, sleep}, @timeout)
  end

  @doc """
  Calls the process and returns the number of calls received by the GenServer and returns the number
  of calls received prior to this one.

  After returning it sleeps for `sleep` milliseconds. Unless the GenServer is otherwise blocked, it
  should not timeout
  """
  @spec before_you_sleep(integer) :: integer
  def before_you_sleep(sleep) do
    GenServer.call(@name, {:before_you_sleep, sleep}, @timeout)
  end

  def handle_call({:yawn, sleep}, _from, call_count) do
    :timer.sleep(sleep)
    {:reply, {:previous_call_count, call_count}, call_count + 1}
  end

  def handle_call({:before_you_sleep, sleep}, from, call_count) do
    GenServer.reply(from, {:previous_call_count, call_count})
    :timer.sleep(sleep)
    {:noreply, call_count + 1}
  end
end
